### ISSUE TITLE
please do not use glyph fonts, icon fonts, or any other abuse of Unicode "Private Use Area" requiring downloaded fonts.

### ISSUE THREAD

**Issue by robohack** [wet-boew/wet-boew#8024]
I have been asked to submit this issue after reporting accessibility issues on the Environment Canada Weather Office and Water Office web pages, since they consider this problem to be an upstream result of having made use of the WET templates.

Your toolkit abuses (i.e. makes non-standard use of) the Unicode "Private Use Area" for what are known as "icon fonts", or "glyph fonts".

When these so-called "icon fonts" are blocked due to security issues, and also when they are not loaded and executed because of browser accessibility features, the web pages using this toolkit become a very confusing and difficult to use.

Many of us refuse to download fonts from the web as they are very often a source of untrusted code and we do not (yet) have the tools to tell the good from the bad.  There are also several popular web browser extensions with commonly used features for blocking remote fonts for this very reason (e.g. uBlock Origin in Chrome).

Furthermore web browsers supporting various accessibility features will not be able to make any sense of icon fonts.

Please refrain from using custom fonts, and in particular you should entirely avoid using so-called "icon fonts" (i.e. custom glyphs in the Unicode private use area).

Good old icon image files are strongly preferred for all buttons and decorative elements on web pages!  They are more easily cached, they can be more easily tagged with alternate descriptive text, and they require fewer resources on both the clients and in most cases on the server too.

You can see how bad things look when custom fonts are blocked by installing the "uBlock Origin" spam blocking extension into Chrome and then configuring it to block all fonts.

Click on uBlock's shield icon then click on the "A" in the bottom right of the menu to "Toggle the blocking of remote fonts for this site" until it shows a red "X" overlaying the "a", then force-reload the page -- you should see empty box icons where you would have expected icon images to appear.

In general users MUST be given the full freedom to choose their own fonts for my own visual preferences and needs.  All too many web designers love to think they're better than anyone else, including all other designers like themselves, at deciding what looks best and then try to force their choices on everyone.  Such autocracy never goes well, especially not with the choice of fonts and specific font-sizes.  Although I personally do not have vision problems, if I did then forcing me to use fonts not suitable to my needs would entirely prevent me from viewing such a site.  All modern web browsers allow their users to choose the best fonts for their display devices and for their personal needs -- web designers must not try to override these choices.

As for the particular problem of icon fonts, well they abuse the Unicode standard.  "Private Use Area" code points should NEVER EVER be seen on any public network and MUST NOT EVER be used between unrelated applications.  They are for internal use only inside a single application that has total control over both data containing these code points and all forms of the presentation of that data.

**Comment by duboisp** [comment]
@robohack Can I reproduce that issue from any pages that use the web experience toolkit or is it only specific to when a author use a [icons](http://wet-boew.github.io/wet-boew-styleguide/v4/design/icons-en.html) like [Glyphicons ](http://getbootstrap.com/components/#glyphicons) in the main content area?

**Comment by robohack** [comment]
Any page will do provided it presents some private-use code points, or makes use of javascript which will replace something in the DOM with a private-use code point, or which makes use of CSS that uses remotely loaded fonts which cover private-use code points.

weather.gc.ca seems to have quite a few, for example.

In the HTML on that page I see these as "span" blocks with a class such as "expicon glyphicon glyphicon-chevron-down", and the related CSS uses a font-family of "Glyphicons Halflings".  These show up as empty boxes when the remote font is blocked.

Also, for example, if I visit glyphicons.com I see nothing but empty boxes in their charts.  Which is what I want.

I don't trust remotely loaded fonts even if they come from the same origin as many web programmers haven't got the skills to audit foreign code, and they commonly download and upgrade such foreign code without any regard to security issues.  (Similarly I block most third-party scripts, and sometimes even same-origin scripts.  That means I block scripts from being loaded from cdn.canada.ca too, for example when visiting any gc.ca site -- cross-domain resources are major security risks.  Nobody outside of gc.ca can knows that gc.ca sites trust canada.ca sites, and assuming blind third-party trust is stupid.  Make sure all resources are always loaded from sites within the same parent domain!)

That's just my security background speaking, though Microsoft themselves suggest blocking untrusted fonts.  https://docs.microsoft.com/en-us/windows/threat-protection/block-untrusted-fonts-in-enterprise

I think your main motivation for avoiding remote fonts should be with the accessibility issues created by abuse of Unicode private-use code points.

**Comment by robohack** [comment]
Some further rationale:

[https://cloudfour.com/thinks/seriously-dont-use-icon-fonts/](Seriously, Don't Use Icon Fonts)

**Comment by duboisp** [comment]
Thanks @robohack for the clarification.

For security, there is a few strategy that can be used, like the [Subresource Integrity](https://www.w3.org/TR/SRI/), [Cross-Origin Resource Sharing](https://www.w3.org/TR/cors/),... which they might require a server side configuration.

**Comment by RobJohnston** [comment]
I have more clarification.  Using @robohack's instructions to block remote fonts for the site, uBlock Origin's logger is triggering the "no-remote-fonts" rule to block both glyphicons-halflings-regular.ttf and glyphicons-halflings-regular.woff.

Is there the possibility of using [glyphicons-halflings-regular.svg](https://github.com/twbs/bootstrap/blob/master/fonts/glyphicons-halflings-regular.svg)?

Not sure how wide-spread this problem is, because one must first go to the site to block the fonts, meaning they already loaded once.

**Comment by LaurentGoderre** [comment]
It's the browser's job to sandbox content downloaded from the Web. Keeping a browser up to date is a more efficient way to deal with security than to block whole content types.

**Comment by robohack** [comment]
Well, I would say that a) sandboxing within an application or language runtime has never proven to be secure, and especially not in something as overly complex and bloated as a modern web browser (even chrome's attempt at using multiple processes has been found to be severely lacking for many types of vulnerabilities); and b) fonts are a special case where security issues are far more subtle since they directly affect the presentation of visual information to the user and in ways that the application has absolutely zero control over.

So, in fact blocking content is the only way to guarantee any real level of security.

**Comment by LaurentGoderre** [comment]
If a user choose to block fonts, they can create a custom stylesheet to display the hidden text that is used to accompany the icon.

**Comment by RobJohnston** [comment]
By uBlock Origin's own issue, ["no-remote-fonts option not working"](https://github.com/gorhill/uBlock/issues/1539#issuecomment-206230700), the use of inline data: URI-based fonts would be a way around this.  That could be a fix for a problem that's been seen here before in https://github.com/wet-boew/wet-boew/issues/5693

**Comment by robohack** [comment]
Don't try to be fancy -- just keep things simple.  Real images as icons are ideal.  SVG icons are probably OK (hopefully soon browsers will make them more secure).

I suspect it is this toolkit which is also responsible for abuse of CPU resources in Chrome sometimes on what should be static wateroffice.gc.ca pages, even when the tab isn't visible.  At best it chews 1% CPU constantly jumping to as much as 4% when the tab is visible but without input focus, and sometimes I see as much as 14% CPU use sustained over hours when the tab isn't visible, and there's a constant growth in memory requirements for the tab too.

**Comment by LaurentGoderre** [comment]
Real image are not ideal because they don't adapt to different pixel density. The Ideal solution is SVG but there are still a few browser bugs preventing us to switch to that.

**Comment by robohack** [comment]
What I mean by "ideal" is something that works everywhere and is most useful to those with accessibility needs -- something icon fonts clearly are the opposite of.

That means:  use real images.

Real images do scale and can compensate for most common pixel densities, and though all scalers are not equivalent, good graphic design can compensate for the most visible problems.

**Comment by LaurentGoderre** [comment]
Images don't come close to SVG  when it comes to rendering icons, especially for high pixel density displays. IN any case, font icons should use aria-hiden to ensure they are not being read out and be accompanied with hidden text to convey the meaning of the icon.

**Comment by RobJohnston** [comment]
With regards to the use of aria-hidden, that is the accessibility advice from both [Glyphicons > How to use > Accessible icons](http://getbootstrap.com/components/#glyphicons-how-to-use) and [Using Font Awesome with Acessibility in mind](http://fontawesome.io/accessibility/), but it doesn't seem to be implemented in WET and isn't in the[ style guide for icons](http://wet-boew.github.io/wet-boew-styleguide/v4/design/icons-en.html).

**Comment by LaurentGoderre** [comment]
I think the recommendation came after the first version of v4 was release. WE can add it it to the example but it's a markup change so it's not guaranteed that any of the implementers would add it to their site.

**Comment by robohack** [comment]
Font icons are still invisible to sighted users when insecure foreign fonts are blocked.  I.e. they make a web page (ab)using them confusing and ugly.

Don't you guys understand the real-world security implications of untrusted fonts!?!?!?

**Comment by duboisp** [comment]
@robohack we are trying to find a solution that will work even when the fonts are blocked. What are you suggesting?

If for a given context the use of icon is "confusing", that might represent an issue to the [WCAG 2.0 Success Criterion 1.1.1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/text-equiv-all.html). In order to prevent it that kind of issue, a solution could be to have clearer instruction and example on how to code them in a web page. Which is @LaurentGoderre suggestion and that I support.

Sorry but being "ugly" do not make a page being non WCAG Level AA compliant.

You state that may be the "fonts" are responsible for an abuse of CPU resources. Is when the font is blocked the CPU usage drop?

Is the security issue that you are referring is only limited to "font-processing-related attacks"? Can you list other vulnerability that is still current and directly related to custom font? Is the current general available release of major browser are still affected by those kind of vulnerability? Is there a way that we can know that a font file we use do not contain that kind of vulnerability?

**Comment by RobJohnston** [comment]
A possible fix for being ugly:  [A Font Garde](https://github.com/filamentgroup/a-font-garde), "A set of reliable (nay, bulletproof) patterns for icon fonts".

With the Unicode substitutions [Left-Pointing Magnifying Glass](http://www.fileformat.info/info/unicode/char/1f50d/index.htm) for the search icon, [Modifier Letter Down Arrowhead](http://www.fileformat.info/info/unicode/char/02c5/index.htm) for the megamenu, and [Heavy Triangle-Headed Rightwards Arrow](http://www.fileformat.info/info/unicode/char/279e/index.htm) for the breadcrumb.  

Don't know what would be good as a fallback for the 'Share' icon, other than maybe [Wrapped Present](http://www.fileformat.info/info/unicode/char/1f381/index.htm).

Thoughts?

**Comment by robohack** [comment]
When I say "ugly" I refer to the square empty boxes that litter the page when foreign fonts are not loaded.

When I say "confusing" I refer to the fact those square empty boxes are supposed to be buttons (even though there's often not even any visual hint that they're supposed to be a button), but there's all too often no alternate text to describe what they do either, and not even any hover text boxes to help.

When I say foreign fonts are insecure, I refer to the fact that the browser (and any foreign code running in the browser) cannot know what glyph is being displayed and seen by the user.  Unicode fonts offer enough potential for similar looking glyphs causing obfuscation, confusion, and fraud, without also allowing entirely untrusted fonts to be used.

If you don't think you can afford (compatability-wise) to use SVG for icons (and/or you can't figure out how to make them work when scripting is disabled or script files won't load), then use images with alt-text!   Images with alt-text ALWAYS work.  With appropriate graphic design and with scaling driven by size restrictions you should be able to get decent enough looking image icons on any modern platform.  Besides, users don't really care if an icon looks a bit "off" -- they really care if they can figure out what will happen if they click on it though.

Don't try to impress your designer friends -- try to impress your users.

**Comment by jfksdca** [comment]
I've done a lot of research on the accessibility of icon fonts.  There are established practices dating back to 2015 which make them perfectly accessible.  By that, I mean that every screen reader knows what to say if they are presented with a custom font that has the appropriate tags.  I agree that we won't be moving backward to use images for icons.  SVGs are beautifully rendered but I don't believe they're perfect.  SVG files increase server requests.  If they're embedded in a page then their code gets loaded over and over again each time the page is requested.  Given that SVG files are a series of text tags I wouldn't be surprised if at some point they were exploited to deliver malicious code.  A custom font is basically a series of SVG files so I don't understand why one is good and the other evil.  Nothing is 100% secure.
